server/lib/two-factor-authentication/webauthn.ts

```diff
@@ -0,0 +1,189 @@
+import * as simplewebauthn from '@simplewebauthn/server';
+import type {
+  AuthenticationResponseJSON,
+  PublicKeyCredentialCreationOptionsJSON,
+  PublicKeyCredentialDescriptorFuture,
+  RegistrationResponseJSON,
+  // eslint-disable-next-line node/no-missing-import, node/no-unpublished-import
+} from '@simplewebauthn/typescript-types';
+import config from 'config';
+
+import { ApolloError } from '../../graphql/errors';
+import { idEncode, IDENTIFIER_TYPES } from '../../graphql/v2/identifiers';
+import User from '../../models/User';
+import UserTwoFactorMethod from '../../models/UserTwoFactorMethod';
+import { TOKEN_EXPIRATION_2FA } from '../auth';
+import cache from '../cache';
+
+import { Token } from './lib';
+import { TwoFactorMethod } from './two-factor-methods';
+
+const WebAuthnTimeoutSeconds = TOKEN_EXPIRATION_2FA;
+const SupportedPublicKeyAlgorithmIDs = [
+  -7, // ES256
+  -8, // EdDSA
+  -257, // RS256
+];
+
+export default {
+  async validateToken(user: User, token: Token, req): Promise<void> {
+    const authenticationResponse: AuthenticationResponseJSON = JSON.parse(
+      Buffer.from(token.code, 'base64').toString('utf-8'),
+    );
+
+    const method = await UserTwoFactorMethod.findOne<UserTwoFactorMethod<TwoFactorMethod.WEBAUTHN>>({
+      where: {
+        UserId: user.id,
+        method: TwoFactorMethod.WEBAUTHN,
+        data: {
+          credentialId: authenticationResponse.id,
+        },
+      },
+    });
+
+    if (!method) {
+      throw new ApolloError('Two-factor authentication code is invalid', 'INVALID_2FA_CODE');
+    }
+
+    await verifyAuthenticationResponse(user, authenticationResponse, req);
+  },
+  async authenticationOptions(user: User, req) {
+    return generateAuthenticationOptions(user, req);
+  },
+};
```

Comment:
It is better to export named functions due to object reference.

```suggestion
export async function validateToken(user: User, token: Token, req): Promise<void> {
  const authenticationResponse: AuthenticationResponseJSON = JSON.parse(
    Buffer.from(token.code, 'base64').toString('utf-8'),
  );
  
  const method = await UserTwoFactorMethod.findOne<UserTwoFactorMethod<TwoFactorMethod.WEBAUTHN>>({
    where: {
      UserId: user.id,
      method: TwoFactorMethod.WEBAUTHN,
      data: {
        credentialId: authenticationResponse.id,
      },
    },
  });
  
  if (!method) {
    throw new ApolloError('Two-factor authentication code is invalid', 'INVALID_2FA_CODE');
  }
  
  await verifyAuthenticationResponse(user, authenticationResponse, req);
}

export async function authenticationOptions(user: User, req) {
  return generateAuthenticationOptions(user, req);
}
```